src/lib.rs

```diff
@@ -284,30 +285,101 @@ impl Authorization {
                     status: AuthorizationStatus::Deactivated,
                 }),
                 self.nonce.take(),
-                &self.url,
+                self.url,
             )
             .await?;
 
-        self.nonce = nonce_from_response(&rsp);
-        self.state = Problem::check::<AuthorizationState>(rsp).await?;
+        *self.nonce = nonce_from_response(&rsp);
+        *self.state = Problem::check::<AuthorizationState>(rsp).await?;
         match self.state.status {
-            AuthorizationStatus::Deactivated => Ok(&self.state),
+            AuthorizationStatus::Deactivated => Ok(self.state),
             _ => Err(Error::Other(
                 "authorization was not deactivated by ACME server".into(),
             )),
         }
     }
 
-    /// Get the [`AuthorizationState`] of the authorization
+    /// Get a [`ChallengeHandle`] for the given `type`
     ///
-    /// Call `refresh()` to get the latest state from the server.
-    pub fn state(&self) -> &AuthorizationState {
-        &self.state
+    /// Yields an object to interact with the challenge for the given type, if available.
+    pub fn challenge(&'a mut self, r#type: ChallengeType) -> Option<ChallengeHandle<'a>> {
+        let challenge = self.state.challenges.iter().find(|c| c.r#type == r#type)?;
+        Some(ChallengeHandle {
+            identifier: &self.state.identifier,
+            challenge,
+            nonce: self.nonce,
+            account: self.account,
+        })
     }
 
     /// Get the URL of the authorization
     pub fn url(&self) -> &str {
-        &self.url
+        self.url
+    }
+}
+
+impl Deref for AuthorizationHandle<'_> {
+    type Target = AuthorizationState;
+
+    fn deref(&self) -> &Self::Target {
+        self.state
+    }
+}
+
+/// A [`Challenge`] wrapper that exposes relevant API
+///
+/// For each challenge, you'll need to:
+///
+/// * Obtain the [`ChallengeHandle::key_authorization()`] for the challenge response
+/// * Set up the challenge response in your infrastructure
+/// * Call [`ChallengeHandle::set_ready()`] for that challenge
+///
+/// After the challenges have been set up, call [`Order::poll()`] to wait until the order is
+/// ready to be finalized (or if it becomes invalid). Once it is ready, call [`Order::finalize()`]
+/// to get the certificate.
+///
+/// Dereferences to the underlying [`Challenge`] for easy access to the challenge's state.
+pub struct ChallengeHandle<'a> {
+    identifier: &'a Identifier,
+    challenge: &'a Challenge,
+    nonce: &'a mut Option<String>,
+    account: &'a AccountInner,
+}
+
+impl ChallengeHandle<'_> {
+    /// Notify the server that the given challenge is ready to be completed
+    ///
+    /// `challenge_url` should be the `Challenge::url` field.
```

Comment:
I think this bit about a `challenge_url` in the docstring is stale now right? No arguments are being passed in by the caller.